packages/store/src/internal/state-context-factory.ts

```diff
@@ -17,44 +17,52 @@ export class StateContextFactory {
   /**
    * Create the state context
    */
-  createStateContext(metadata: MappedStore): StateContext<any> {
+  createStateContext<T>(metadata: MappedStore): StateContext<T> {
     const root = this._internalStateOperations.getRootStateOperations();
     return {
-      getState(): any {
+      getState(): T {
         const state = root.getState();
         return getValue(state, metadata.depth);
       },
-      patchState(val: any): any {
-        const isArray = Array.isArray(val);
-        const isPrimitive = typeof val !== 'object';
-
-        if (isArray) {
-          throw new Error('Patching arrays is not supported.');
-        } else if (isPrimitive) {
-          throw new Error('Patching primitives is not supported.');
-        }
+      patchState(val: Partial<T> | PatchFunction<T>): T {
+        const isFunction = typeof val === 'function';
+        const patchOperation = isFunction ? <PatchFunction<T>>val : simplePatch<T>(<Partial<T>>val);
 
         const state = root.getState();
         const local = getValue(state, metadata.depth);
-        const clone = { ...local };
-
-        for (const k in val) {
-          clone[k] = val[k];
-        }
+        const clone = patchOperation(local);
 
         const newState = setValue(state, metadata.depth, clone);
         root.setState(newState);
         return newState;
       },
-      setState(val: any): any {
+      setState(val: T): T {
```

Comment:
I think PatchFunction it will be the best feature

```ts
setState(val: T | PatchFunction<T>): T {
  let state = root.getState();
  const stateValueOperation = this.getValueByOperation(val, CTX_CONTEXT.SET_VALUE); 
  const stateValue = stateValueOperation(state); // (state) => { ...state }
  state = setValue(state, metadata.depth, stateValue);
  ...
}
```

Before

```ts
@Action(MyAction)
public addValue(ctx: StateContext, { payload }: MyAction) {
  ctx.setState({ ...ctx.getState(), value: payload  });
}
```

After

```ts
@Action(MyAction)
public addValue(ctx: StateContext, { payload }: MyAction) {
  ctx.setState((state) => ({ ...state, value: payload  }));
}
```

```ts
patchState(val: Partial<T> | PatchFunction<T>): T {
  const state = root.getState();
  const local = getValue(state, metadata.depth);
  const stateValueOperation = this.getValueByOperation(val, CTX_CONTEXT.PATCH_STATE); 
  const stateValue = stateValueOperation(local); // this.simplePatch(val)(local);
  const newState = setValue(state, metadata.depth, stateValue);
}
```

```ts
@Action(MyAction)
public addValue(ctx: StateContext, { payload }: MyAction) {
  // state = ['1', '2'], payload = '3'
  ctx.patchState(() => ([ payload ])); // state === ['1', '2', '3']
}
```

Reply:
So, are you saying that it would be better on the `setState` function as opposed to the `patchState` function?
I think I like that. Then the `patchState` function could remain unchanged.

Reply:
It would be logical to have the same interface for setState, patchState

Reply:
I think that this makes so much more sense for `setState`. 
Having this option for the `patchState` function could leave some ambiguity... ie. what should I be returning from the function? The new state or the value to be used for the patch?

Reply:
setState -> return new state
patchState -> value for state

Reply:
I don't really see any value in this one:
`patchState -> value for state`
I'll just do the one